event/handler.go

```diff
@@ -0,0 +1,108 @@
+package event
+
+import (
+	"time"
+	"github.com/ONSdigital/go-ns/log"
+	"github.com/ONSdigital/dp-import-reporter/model"
+)
+
+//go:generate moq -out ../mocks/event_generated_mocks.go -pkg mocks . DatasetAPICli Cache EventHandler
+
+const (
+	failed                    = "failed"
+	errorType                 = "error"
+	datasetAPIGetInstErr      = "datasetAPI.GetInstance return an error"
+	datasetAPIAddEventErr     = "datasetAPI.AddEventToInstance return an error"
+	datasetAPIUpdateStatusErr = "datasetAPI.UpdateInstanceStatus return an error"
+	eventNotInCache           = "report event not found in cache, checking dataset API"
+	eventNotInInstance        = "report not found in instance.events, updating dataset api"
+	updatingDSInstance        = "updating dataset api to set instance.status to failed"
+	addingToLocalCache        = "adding report event to local cache"
+	updatingCacheTimeout      = "report event found in cache, updating cache expiry time"
+	handlingEvent             = "Handling report event"
+	reportEventKey            = "reportEvent"
+	generateCacheKeyValueErr  = "error while attempting to generate cache key value for report event"
+)
+
+var (
+	statusFailed = &model.State{State: failed}
+)
+
+// DatasetAPICli defines that interface for a client of the DatasetAPI
+type DatasetAPICli interface {
+	GetInstance(instanceID string) (*model.Instance, error)
+	AddEventToInstance(instanceID string, e *model.Event) error
+	UpdateInstanceStatus(instanceID string, state *model.State) error
+}
+
+// Cache defines the behaviour of an in memory cache
+type Cache interface {
+	Get(key []byte) (value []byte, err error)
+	Set(key, value []byte, expireSeconds int) (err error)
+	Del(key []byte) (affected bool)
+	TTL(key []byte) (timeLeft uint32, err error)
+}
+
+// Handler a struct for handling a ReportEvent
+type Handler struct {
+	DatasetAPI    DatasetAPICli
+	Cache         Cache
+	ExpireSeconds int
+}
+
+// HandleEvent handle the reportEvent
+func (r Handler) HandleEvent(e *model.ReportEvent) error {
```

Comment:
Just realised that there are some cases where we will not have an instanceID - what can we do in this scenario? We can't update the job because... which one do we update?

Reply:
I suggest handle a case for such an event by logging that it has happened and continue reading events. Meanwhile any service writing to report event topic should not be sending events which do not contain any relevant information.

In the future there maybe scenarios whereby we want to handle an event with no instance id because it has other valuable information.